text/0000-introduce empty ranges.md

```diff
@@ -0,0 +1,425 @@
+>           or _max.nan() or _max.infinite()
+>           or _inc.nan() or _inc.infinite()
+>       else
+>         false
+>       end
+>     _infinite =
+>       is_float_infinite
+>         or ((_inc == 0) and (min != max))    // no progress
+>         or ((_min < _max) and (_inc < 0)) // progress into other directions
+>         or ((_min > _max) and (_inc > 0)) 
+> ```
+
+This RFC partitions the set of currently *infinite* `Range` cases into those now considered *empty*, and a very small set of those that remain *infinite*.
+
+For this, two criteria are postulated that are required for a Range to not be empty. Numeric values including the floating point values `+Inf` and `-Inf` can be meaningfully compared in expressions like `min > max` or `max > min`. Based on this, an obvious absence of *progress* of the iteration from min towards max based on the sign of the Range parameters can be tested by `((min <= max) and (step <= 0)) or ((min >= _max) and (step >= 0))` ("no-progress" criterion) for finite and infinite bounds as well as finite and infinite values of step (note that if either of the `min` or `max` parameters is `NaN` such expressions are always `false` and this special case is discussed further below). Here, besides the obvious sign-incompatibility of the three Range parameters, the cases of equal bounds `min == max` and `step == 0` are included, since the former wouldn't satisfy `[min, min)` (and also doesn't require any progress) and the latter prevents any progress. In the current implementation of Range, a somewhat similar couple of expressions are used with the comment `// no progress` and `// progress into other directions`, but they are treated as a sufficient condition for *infinite*. Here, this criterion is used as part of a test for empty Ranges because it allows us to know in which cases the generation of values between min and max is impossible, irrespective of the magnitude of the bounds or steps, in other words, no valid *trajectory* from min to max does exists with the provided parameters. The second important criterion used here is that for a Range iterators to be not empty, it must only produce finite values even when infinite bounds are specified, and if that is not possible, the Range is again *empty*.
+
+Combining these two criteria, for a `Range` to not be *empty*, progress from `min` to `max` must be possible (no-progess is `false`), and the iterator that realizes this progress must produce *finite* values that lie within `[min, max)`.
+
+Using this test, `Range(0, 10, -1)` is *empty* and no longer *infinite*, but so is in fact `Range(0, inf, -1)` that is also considered *infinite* currently. While one can generally never produce a complete list of elements that would incrementally move from say 0 to +Inf, we can know in the case of `Range(0, inf, -1)` that the Range must be *empty* because of the incompatible sign of the step parameter relative to the signs of `min` and `max` (progress impossible).
```

Comment:
The argument against transforming `Range(0, 10, -1)` into `Range(10, 0, -1)` is that by the stated parameters, Range evaluates those numbers using *expressions*. The underlying idea is that just like an if clause prevents a block of code to be executed if the condition is not met, the Range should be empty if those conditions are not met. Re-arranging/swapping the parameters that feed into the underlying expression and thereby basically changing the meaning of the parameters `min` and `max` is then no weirder than swapping the arguments in an if clause so that the code block can be executed.

Reply:
I see what you mean, but still do not think you have made a case of "empty" ranges over "parameter rearrangement" as a solution. I put more thought into this and the critical element is whether we want to keep an implied directionality to `min` and `max` which then interacts with the direction of `step` or if directionality should *only* be determined by `step` with `min` and `max` simply setting the interval.

In the latter, `Range(-1, -11, -1)` would indeed iterate from -1 to -11, while `Range(-1, -11, 1)` would iterate from -11 to -1. I think the biggest argument against a parameter rearrangement is the half-closed interval of range, which should not change. If I provide two values (`min` and `max` as the are named now), I should know which "side" it will stop at. Empty ranges approach this better, but there might be a good use case for some `Interval` class which works via parameter rearrangement which is *not* half-closed. Changing the half-closed nature of Range is far too breaking of a change for parameter rearrangement to be viable.

Once you have read this message, please resolve this conversation thread. I am sufficiently on the side of empty range over parameter rearrangement now. It only took me sitting on the idea for a day to realize the problem with parameter rearrangement.